src/compiler/scala/tools/nsc/transform/CleanUp.scala

```diff
@@ -383,6 +380,106 @@ abstract class CleanUp extends Statics with Transform with ast.TreeDSL {
       }
     }
 
+    object StringsPattern {
+      def unapply(arg: Tree): Option[List[String]] = arg match {
+        case Literal(Constant(value: String)) => Some(value :: Nil)
+        case Literal(Constant(null))          => Some(null :: Nil)
+        case Alternative(alts)                => traverseOpt(alts)(unapply).map(_.flatten)
+        case _                                => None
+      }
+    }
+
+    // transform scrutinee of all matches to ints
+    def transformSwitch(sw: Match): Tree = { import CODE._
+      sw.selector.tpe match {
+        case IntTpe => sw // can switch directly on ints
+        case StringTpe =>
+          // these assumptions about the shape of the tree are justified by the codegen in MatchOptimization
+          val Match(Typed(selTree: Ident, _), cases) = sw
+          val sel = selTree.symbol
+          val restpe = sw.tpe
+          val swPos = sw.pos.focus
+
+          /* From this:
+           *     string match { case "AaAa" => 1 case "BBBB" | "c" => 2 case _ => 3}
+           * Generate this:
+           *     goto matchSuccess (string.## match {
+           *       case 2031744 =>
+           *         if ("AaAa" equals string) goto match1
+           *         else if ("BBBB" equals string) goto match2
+           *         else goto matchFailure
+           *       case 99 =>
+           *         if ("c" equals string) goto match2
+           *         else goto matchFailure
+           *       case _ => goto matchFailure
+           *     }
+           *     match1: goto matchSuccess (1)
+           *     match2: goto matchSuccess (2)
+           *     matchFailure: goto matchSuccess (3) // would be throw new MatchError(string) if no default was given
+           *     matchSuccess(res: Int): res
+           * This proliferation of labels is needed to handle alternative patterns, since multiple branches in the
+           * resulting switch may need to correspond to a single case body.
+           */
+
+          val stats = mutable.ListBuffer.empty[Tree]
+          var failureBody = Throw(New(definitions.MatchErrorClass.tpe_*, REF(sel))) : Tree
```

Comment:
I can't remember why I thought I couldn't use the synth default case from patmat. This should never be used (the `Ident(WILDCARD, _, body)` case should always be hit but it seems safer.